src/Stream/src/Base/Binder/EmbeddedHeaderUtils.cs

```diff
@@ -0,0 +1,179 @@
+                    {
+                        byteBuffer.WriteByte((byte)headers[i].Length);
+
+                        var buffer = Encoding.UTF8.GetBytes(headers[i]);
+                        byteBuffer.Write(buffer, 0, buffer.Length);
+
+                        buffer = GetBigEndianBytes(headerValues[i].Length);
+                        byteBuffer.Write(buffer, 0, buffer.Length);
+
+                        byteBuffer.Write(headerValues[i], 0, headerValues[i].Length);
+                    }
+                }
+
+                var payloadBuffer = (byte[])original.Payload;
+                byteBuffer.Write(payloadBuffer, 0, payloadBuffer.Length);
+
+                return byteBuffer.ToArray();
+            }
+            catch (Exception)
+            {
+                // Log
+                throw;
+            }
+        }
+
+        public static bool MayHaveEmbeddedHeaders(byte[] bytes)
+        {
+            return bytes.Length > 8 && (bytes[0] & 0xff) == 0xff;
+        }
+
+        public static MessageValues ExtractHeaders(byte[] payload)
+        {
+            return ExtractHeaders(payload, false, null);
+        }
+
+        public static MessageValues ExtractHeaders(IMessage<byte[]> message, bool copyRequestHeaders)
+        {
+            return ExtractHeaders(message.Payload, copyRequestHeaders, message.Headers);
+        }
+
+        public static string[] HeadersToEmbed(string[] configuredHeaders)
+        {
+            if (configuredHeaders == null || configuredHeaders.Length == 0)
+            {
+                return BinderHeaders.STANDARD_HEADERS;
+            }
+            else
+            {
+                var combinedHeadersToMap = new List<string>(BinderHeaders.STANDARD_HEADERS);
+                combinedHeadersToMap.AddRange(configuredHeaders);
+                return combinedHeadersToMap.ToArray();
+            }
+        }
+
+        private static MessageValues ExtractHeaders(byte[] payload, bool copyRequestHeaders, IMessageHeaders requestHeaders)
+        {
+            var byteBuffer = new MemoryStream(payload);
+            var headerCount = byteBuffer.ReadByte() & 0xff;
+            if (headerCount == 0xff)
+            {
+                headerCount = byteBuffer.ReadByte() & 0xff;
+                var headers = new Dictionary<string, object>();
+                for (var i = 0; i < headerCount; i++)
+                {
+                    var len = byteBuffer.ReadByte() & 0xff;
+                    var headerName = Encoding.UTF8.GetString(payload, (int)byteBuffer.Position, len);
+
+                    byteBuffer.Position += len;
+
+                    var intBytes = new byte[4];
+                    byteBuffer.Read(intBytes, 0, 4);
+                    len = GetIntFromBigEndianBytes(intBytes);
+                    var headerValue = Encoding.UTF8.GetString(payload, (int)byteBuffer.Position, len);
+                    var headerContent = JsonConvert.DeserializeObject(headerValue);
```

Comment:
```suggestion
                    var headerContent = JsonSerializer.Deserialize<string>(headerValue, new JsonSerializerOptions { Encoder = JavaScriptEncoder.Create(UnicodeRanges.BasicLatin, UnicodeRanges.All) });
```

Reply:
This code change looks like you are assuming that the header content is always a string... is that correct? 
If so, that won't work as whats embedded is a stream of bytes that contain JSON objects and those can end up not strings (e.g. long=> timestamp, Guid=>correlationid, etc.)

Reply:
Ah, yeah I possibly oversimplified it then... On mobile now, but I can update the suggestion next week if you don't have time to figure out the non-Newtonsoft way to do this. FWIW, all the tests pass this way

Reply:
I'll update a test to make sure it validates its getting the correct types.